# הצעת מחקר לסדנה במדעי הנתונים - עומר חפץ

### פרטי המגיש

שם: עומר חפץ

ת.ז: 208786244

### מטרת הפרויקט

מערכות המלצה הן בעלות פוטנציאל גדול לקדם את המכירות של עסקים ממגוון סוגים. עם זאת, כל מסד נתונים של חברה הוא ייחודי, ודורש התאמות כדי שיגיע לביצועים טובים

הבעיה אותה אני מעוניין לפתור היא בהינתן דאטה איכותי של חברה, לבנות את מערכת ההמלצה הטובה ביותר בהתבסס על מגוון השיטות הקיימות ועל מחקר של מימוש מערכות המלצה באזרחות 

במילים אחרות, להנגיש למשתמש בלחיצת כפתור את מודל ההמלצות המוצלח ביותר, וכמובן גם את הביצועים שלו

### מוטיבציה

מערכות המלצה הן בעלות ערך אדיר לשוק הפרטי, ומימוש טוב שלהן יכול להביא לפוטנציאל רווחי משמעותי כפי שאנו רואים בחברות הגדולות (נטפליקס, אמזון, ופייסבוק)

### הסבר על הדאטה והאתגר הקיים בפרויקט

על המערכת שתיבנה בפרויקט למצוא את המודל המוצלח ביותר ללא כל היכרות עם כמות וסוגי הפיצ'רים שיעמדו לרשותה. 

באופן בסיסי, הדאטה בוודאות יכיל רצפים של דירוגים והמלצות של משתמשים מהשירות - על הימצאות הדאטה זה ניתן לסמוך.

עם זאת, לא ניתן יהיה לדעת מראש את המידע הבא, וזה יהווה האתגר הגדול בפרויקט:
* מהו המטא-דאטה שמוצמד למוצר או למשתמשים, וכיצד נכון להשתמש בו כדי לקבל תוצאות אופטימליות
* שימוש בפיצ'רים מבוססי תוכן (NLP)
* מהי סקאלת הדירוגים האפשרית של המוצר
* כיצד מתפלג הדאטה בין כמות המשתמשים, כמות המוצרים, וכמות הדירוגים
* פיצ'רים ייחודיים נוספים? (זמנים? תמונות? שפה כמצוין למעלה?) - כל אחד מהם ידרוש לבנות ארכיטקטורה מורכבת יותר שתכיל כמה רכיבים מבוססי למידת מכונה שמתקשרים האחד עם השני

### המקור של הדאטה

מקור הדאטה במאגרי נתונים אותם איתרתי באינטרנט, אשר מופיעים בפורמטים שונים. חלקם מופיעים בתוך קבצי טקסט, חלק בג'ייסון וחלק כסי-אס-וי (מתנצל על העברית, לא ניתן לשלב עברית ואנגלית בצורה נוחה במחברת). מטרת הפרויקט תהיה לעבוד עם מספר סטים לפחות, ולמצוא את הפתרונות הטובים ביותר והשימוש היעיל ביותר באלגוריתמים. 

דוגמה לקישור לדאטה סטים רצויים לפרויקט: https://cseweb.ucsd.edu/~jmcauley/datasets.html
כאשר המאגרים הרלוונטיים לפרויקט לדוגמה יכולים להיות:

* ביקורות על מוצרי אמזון - מהדאטה סטים הגדולים ביותר במאגר
* ביקורות על מתכוני בישול
* מאגר של ביקרות על סרטים שנקרא MovieLens שמשמש גם כסטנדרנט לבחינת ביצועי מנוע וישמש עבורנו כהשוואה לשאר הדברים
* ביקורות על סוגי בירות שונות
* וכדומה

## ניסוי עם הנתונים ותוצאות ראשוניות

בניתוח מטרת הפרויקט וקריאת חומרי רקע על עולם הבעיה, נתקלתי בשתי בעיות מעט שונות שמערכות המלצה מתיימרות לפתור:
1. בהינתן משתמש, על המערכת להמליץ על אוסף מוצרים שהם המתאימים ביותר עבורו
2. בהינתן משתמש ומוצר, על המערכת לחזות את הציון (דירוג) שהמשתמש היה נותן לאותו המוצר

בפרויקט זה החלטתי להסתכל על בעיה 1 כתת בעיה של בעיה 2, ועל כן להתמקד בבעיה מס' 2.

---

כלומר, מטרתנו היא לבצע חיזוי בהינתן משתמש ומוצר - ולחזות מהו הציון שהמשתמש היה נותן לאותו המוצר

**נתנסה במשימה זו עם נתונים ראשוניים ונוציא מסקנות ראשוניות**

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

**חידוד הוא כי מטרת הפרויקט להריץ פייפליין מסודר שיוכל להתמודד עם מספר גדול של נתוני אימון מסוגים שונים. נבחר כאן נתונים מסוג אחד לטובת חיכוך עם האלגוריתמים והדאטה**

נעבוד עם סט נתונים שמכיל 100 אלף ביקורות על סרטים על ידי משתמשים.

נבצע תהליך אקספלורציה והיכרות עם הנתונים

קישור:

https://files.grouplens.org/datasets/movielens/ml-latest-small-README.html

loading the data to dataframes

In [2]:
ratings_df = pd.read_csv("data/ml-100k/ratings.csv", encoding="utf-8")

tags_df = pd.read_csv("data/ml-100k/tags.csv", encoding="utf-8")

movies_df = pd.read_csv("data/ml-100k/movies.csv", encoding='utf-8')

visualize the df content

### ratings

In [3]:
ratings_df.head(10)

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931
5       1       70     3.0  964982400
6       1      101     5.0  964980868
7       1      110     4.0  964982176
8       1      151     5.0  964984041
9       1      157     5.0  964984100

In [4]:
print("Rating dataframe size: " + str(ratings_df.shape))

Rating dataframe size: (100836, 4)


נבחן האם ישנם ערכים חסרים:

In [5]:
print("Rating dataframe size without NAN values: " + str(ratings_df.dropna().shape))

Rating dataframe size without NAN values: (100836, 4)


נשים לב כי במקרה שלנו אין ערכים חסרים במטריצה, אז אין צורך לבצע ניקוי

In [6]:
print("Min rating value is: " + str(ratings_df['rating'].max()))
print("Max rating value is: " + str(ratings_df['rating'].min()))

Min rating value is: 5.0
Max rating value is: 0.5


כפי שמוצג למעלה, הערך המינימלי והמקסימלי של הדירוג תואם את טווח הציונים האפשרי שתואר בסט הנתונים, ולכן נראה שגם בסעיף זה אין צורך לבצע ניקוי יסודי של הנתונים

In [7]:
ratings_df.describe()

userId        movieId         rating     timestamp
count  100836.000000  100836.000000  100836.000000  1.008360e+05
mean      326.127564   19435.295718       3.501557  1.205946e+09
std       182.618491   35530.987199       1.042529  2.162610e+08
min         1.000000       1.000000       0.500000  8.281246e+08
25%       177.000000    1199.000000       3.000000  1.019124e+09
50%       325.000000    2991.000000       3.500000  1.186087e+09
75%       477.000000    8122.000000       4.000000  1.435994e+09
max       610.000000  193609.000000       5.000000  1.537799e+09

נשים לב שמעניין לראות שציון הדירוג הממוצע הוא 3.5, בעוד המינימלי הוא 0.5 והמקסימלי הוא 5. כלומר, היינו מצפים לציון ממוצע של 2.75 אם הנתונים היו מתפלגים בצורה אחידה או נורמלית בין כל הדירוגים האפשריים. 

אין הרבה משמעות לנתונים על סטיית התקן והערך הממוצע של הסרטים והמשתמשים, שכן הם מקבלים ערכים שרירותיים ולא מספר שנשתמש בו לחישוב כלשהו

נסתכל על מספר הסרטים היוניקיים ומספר המשתמשים היוניקיים שקיבלו וביצעו דירוגים בהתאמה:

In [8]:
print("Number of unique users adding ratings: " + str(ratings_df['userId'].nunique()))

Number of unique users adding ratings: 610


In [9]:
print("Number of unique movies with ratings: " + str(ratings_df['movieId'].nunique()))

Number of unique movies with ratings: 9724


ישנם 610 משתמשים שביצעו תיוג, וכן 9724 סרטים. נשמור את מספרים אלו להמשך:

In [10]:
n_movies = ratings_df['movieId'].nunique()
n_users = ratings_df['userId'].nunique()

### tags

In [11]:
tags_df.head(10)

userId  movieId                tag   timestamp
0       2    60756              funny  1445714994
1       2    60756    Highly quotable  1445714996
2       2    60756       will ferrell  1445714992
3       2    89774       Boxing story  1445715207
4       2    89774                MMA  1445715200
5       2    89774          Tom Hardy  1445715205
6       2   106782              drugs  1445715054
7       2   106782  Leonardo DiCaprio  1445715051
8       2   106782    Martin Scorsese  1445715056
9       7    48516       way too long  1169687325

In [12]:
print("Tags dataframe size: " + str(tags_df.shape))

Tags dataframe size: (3683, 4)


נבחן האם קיימים פה ערכים חסרים

In [13]:
print("Tags dataframe size without NAN values: " + str(tags_df.dropna().shape))

Tags dataframe size without NAN values: (3683, 4)


ונראה כפי שרשום למעלה שגם פה אין ערכים חסרים

נבחן את כמות היוזרים היוניקיים (השונים) שנתנו תגית כלשהי לסרט

כמו כן, נבחן את כמות התגיות השונות. למרות שמדובר בטקסט חופשי, אולי נגלה כי היה שימוש חוזר באותן המילים - וזה יעזור לנו לבצע ניתוח של הפעולות האפשריות על הדאטה ופיצ'רים פוטנציאליים לאימון

In [14]:
print("Number of unique users adding tags: " + str(tags_df['userId'].nunique()))
print("Number of unique movies with tags: " + str(tags_df['movieId'].nunique()))
print("Number of unique tags: " + str(tags_df['tag'].nunique()))

Number of unique users adding tags: 58
Number of unique movies with tags: 1572
Number of unique tags: 1589


באופן מפתיע, למרות שמדובר בטקסט חופשי, נשים לב שמתוך 3683 רשומות, רק 1589  מהן יוניקיות, כלומר - פחות ממחצית

בנוסף, נשים לב שכמות הסרטים שקיבלו תיוג היא 1572, ורק על ידי 58 משתמשים.

נחשב את החלק היחסי שלהם מסך כל הסרטים והמשתמשים:

In [15]:
print("Percentage of users tagging movies from all users: " + str(tags_df['userId'].nunique() / n_users))
print("Percentage of tagged movies by users from all movies: " + str(tags_df['movieId'].nunique() / n_movies))

Percentage of users tagging movies from all users: 0.09508196721311475
Percentage of tagged movies by users from all movies: 0.1616618675442205


נשים לב שאחוז הסרטים המתויגים גדול כמעט פי 2 מאחוז המשתמשים המתייגים - ועל כן יתכן שמומלץ יהיה לבחון שימוש בנתון זה כפיצ'ר עבור הסרטים עצמם, ולא עבור המשתמשים המתייגים

### movies

In [16]:
movies_df.head(10)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   
5        6                         Heat (1995)   
6        7                      Sabrina (1995)   
7        8                 Tom and Huck (1995)   
8        9                 Sudden Death (1995)   
9       10                    GoldenEye (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy  
5                        Action|Crime|Thriller  
6                               Comedy|Romance  
7                           Adventure|Children  
8                                       Action  
9                    Action|Adventure|Thriller

In [17]:
print("Movies dataframe size: " + str(movies_df.shape))

Movies dataframe size: (9742, 3)


In [18]:
print("Movies dataframe size without NAN values: " + str(movies_df.dropna().shape))

Movies dataframe size without NAN values: (9742, 3)


גם במטריצה זו אין ערכים חסרים שאיתם עלינו להתמודד

## חלוקת הנתונים ותהליך האימון

נבצע חלוקה בין נתוני אימון לנתוני מבחן בשלב זה. נשים לב שנרצה להשאיר נתוני טסט לכל אחד מהמשתמשים המדרגים, ולכן נשתמש במתודה מיוחדת שהספריה מאפשרת לנו

stratify היכולת

In [19]:
X = ratings_df.copy()
y = ratings_df['userId'] # define as y only for stratifing purpse. then, use the ratings as the real 'y' column

# using stratify to get data for each user
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, stratify=y, random_state=42)

נוודא שהחלוקה בוצעה כמו שרצינו, כלומר, לדוג', שעבור משתמש מס' 1 יופיעו 90 אחוז מהדירוגים שלו באימון, ו-10 אחוזים במבחן

In [20]:
user_1_train_m = X_train[X_train['userId'] == 1].shape[0]
user_1_total_m = X[X['userId'] == 1].shape[0]

user_1_train_m / user_1_total_m * 100

90.08620689655173

וקיבלו 90% כפי שרצינו

In [21]:
# check train-test sizes:
print("Train data shape: " + str(X_train.shape))
print("Test data shape: " + str(X_test.shape))

Train data shape: (90752, 4)
Test data shape: (10084, 4)


y-אין לנו צורך ב

stratify-משום שבפועל הערך הרצוי קיים בתוך הנתונים, כפי שהסברנו למעלה לגבי ה

#### נגדיר שיטת מדידה לביצועי המודלים שלנו

RMSE נשתמש במדד השגיאה הריבועית הממוצעת

הסיבה לכך היא שמדדים כמו דיוק לא מייצגים בצורה טובה מספיק את המציאות. 

**דוגמה להמחשה**: אם הציון הנכון של משתמש על מוצר הוא 2 ומודל ראשון חזה ציון 3 ושני ציון 5, ברור כי המודל הראשון מוצלח יותר, אך במדדי דיוק שני המודלים שגו באותו האופן

RMSEנשתמש בממוצע השגיאות המרובעות, ונבצע עבורו שורש בעצמנו לטובת קבלת ה

In [22]:
# working on lists as well
def rmse(y_true, y_pred):
    return np.sqrt(mean_squared_error(y_true, y_pred))

In [23]:
# example sets:
ratings_true = [1, 5, 2, 4, 1, 3]
ratings_pred = [2, 3, 2, 2, 1, 4.5]

rmse(ratings_true, ratings_pred)

1.3693063937629153

#### evaluating a model on our test data

נגדיר מודל בסיסי שאליו נוכל לבצע השוואה עם המודלים שלנו. מטרתו - להבין כמה שיפור ביצענו בעזרת האלגוריתם. כל מודל שנגדיר יקבל שני פרמטרים - מספר משתמש ומספר מוצר, והוא יצטרך להחזיר כפלט את הציון החזוי לאותו המוצר.

הסיבה שאנו בונים את המודלים בצורה הזו, היא כדי לפתור את בעיה מספר 2 אותה תיארנו למעלה ושאותה אנחנו מעוניינים לפתור. נעתיק אותה בשנית:

**בהינתן משתמש ומוצר, על המערכת לחזות את הציון (דירוג) שהמשתמש היה נותן לאותו המוצר**

---

נבנה מודל בסיס שיחזה את הציון הממוצע לכל סרט מבין הטווח 0.5 ועד 5, שהוא 2.75 

In [24]:
# we will define working models as a function that can get a single user and an item, and produce a rating
# this is exactly problem no.2 defined above

# baseline model for example
def baseline_model(user, item):
    # return middle score between 0.5 and 5
    return 2.75

נגדיר פונ' עזר שמחשבת את הציון של האלגוריתמים שלנו. מה שהיא עושה בפועל זה לקחת את נתוני הטסט שלנו, לבצע חיזוי לכל משתמש וסרט שיש לנו עליו נתון - ואז להשוות RMSE-לציון המקורי שנתן אותו המשתמש, ולגזור מההפרש את ערך ה 

In [25]:
# define a score function for evaluation of the rmse on a model
def score(model):
    
    user_movie_pairs = zip(X_test['userId'], X_test['movieId'])
    
    y_pred = np.array([model(user, movie) for (user, movie) in user_movie_pairs]).reshape(1, -1)
        
    y_true = np.array([X_test['rating']])
    
    return rmse(y_true=y_true, y_pred=y_pred)

לדוג', להלן מימוש של פונקציית הציון על המודל הבסיסי שהגדרנו:

In [26]:
# example of model usage
score(baseline_model) # this is the score that we would like to beat with our models

1.2825975694614493

---

#### Training models

כעת נבצע הרצה של שני אלגוריתמים של מערכות המלצה כדי לבצע השוואה לתוצאות:

1. הראשון מבוסס למידה בלתי מפוקחת של ביצוע קליסטור של המשתמשים לפי הדירוגים שלהם, ומתן הציוג לפי ממוצע מבוסס המשתמשים הדומים להם בלבד לפי תוצאת הקלסיטור
2. השני מבוסס תהליך כפל מטריצות שנועד לחלץ מס' משתנים מצומצם לכל משתמש ולכל מוצר. מבוסס על פתרון לתחרות של נטפליקס משנת 2006 של סימון פאנק

In [27]:
# import the surprise recommender system library
from surprise import KNNBasic, SVD
from surprise.model_selection import cross_validate
from surprise import Reader, Dataset

In [28]:
knn_alg = KNNBasic(k=100, min_k=1)

svd_alg = SVD()

reader נטען את הנתונים לפי מה שדורשת הספרייה - לתוך אובייקט שנקרא

נשים לב כי בשלב זה אנו משתמשים באלגוריתמים אלו אך ורק בעזרת נתוני הדירוג של המשתמשים. אנו לא משתמשים כלל בנתוני מסגרת נוספים שנתקלנו בהם בשלב האקספלורציה במידע. דוגמה לנתונים נוספים שיכולנו להשתמש בהם בתהליך:
1. התיוגים שנתנו המשתמשים על סרטים
2. הז'אנרים של הסרטים השונים

In [29]:
# use a reader to parse the csv dataFrame
reader = Reader(rating_scale=(1, 5))

data = Dataset.load_from_df(X_train[['userId', 'movieId', 'rating']], reader)

---

וכעת, לאחר טעינת הנתונים נבצע הרצות של שני האלגוריתמים בנפרד:

ונתעלם מנתוני המבחן בשלב זה KFOLD נחדד כי בהרצה הנוכחית נשתמש בחלוקת הנתונים ובאימון בשיטת  

#### KNN הרצת

In [30]:
print(cross_validate(knn_alg,
               data=data,
               measures=['rmse'],
               cv=2))

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
{'test_rmse': array([0.98515077, 0.98229019]), 'fit_time': (0.062400102615356445, 0.07800006866455078), 'test_time': (2.2628445625305176, 2.1664211750030518)}


#### SVD הרצת

In [31]:
print(cross_validate(svd_alg,
                     data=data,
                     measures=['rmse'],
                     cv=2))

{'test_rmse': array([0.89293122, 0.89238824]), 'fit_time': (3.2830276489257812, 3.278005599975586), 'test_time': (0.45240092277526855, 0.4478015899658203)}


## ניתוח בסיסי של התוצאות

נשים לב שקיבלנו תוצאות שהלכו והשתפרו עם המודלים שמימשנו:
1. מודל הביסליין ה"רגיל" שלמעשה לא מבצע שום חיזוי חכם למעט הציון הממוצע החזיר לנו שגיאה של 1.28
2. המודל מבוסס הקלסיטור החזיר לנו ממוצע שגיאה של 0.98
3. המודל מבוסס כפל המטריצות החזיר את התוצאה הטובה ביותר, והיא שגיאה ממוצע של 0.89

נשים לב כי הצלחנו להגיע לשיפורים גדולים משימוש בשני אלגוריתמים ראשוניים וזאת אף ללא כל ניצול של כל הנתונים שעמדו לרשותינו. 

# צעדי המשך:
1. לממש אלגוריתמים פוטנציאליים נוספים באמצעות המתודות שמתאפשרות בספרייה איתה עבדנו, ולא לבנות פונקציות נפרדות לכל מודל בנפרד.
2. עם סט נתונים זה לא נדרשנו כמעט בכלל לבצע ניקוי או בדיקה מעמיקה של הנתונים (נתונים חסרים, נתונים מחוץ לטווח, וכו'). יש לממש בהמשך הפרויקט תהליך שעושה ניקוי מסודר לנתונים - משמיט ערכים לא תקינים, משלים/משמיט ערכים חסרים, וכו'.
3. מימוש אלגוריתמים של שיטות היברידיות שמשלבים גם פיצ'רים על המוצרים / המשתמשים עצמם (כדוגמת הז'אנרים של הסרטים או התגיות של המשתמשים בסט הנתונים הנ"ל), ולשלב את הנתונים האלו יחד עם שימוש באלגוריתמים שבוחנים רק את דירוגי המשתמשים.
4. מימוש השוואות מסודרות ובחינת מטריקות נוספות להשוואה בין הנתונים. לדוגמא:

MAE, Maybe Recall + Precision? More matrices for Ordinal Regression

5. חלוקה נוספת של הנתונים גם לסט של ולידיישן. משום שאנחנו רוצים לבחון מודלים שונים ופרמטרים שונים, אנחנו מקבלים מצב שההערכה שלנו על סט הטסט איננה מייצגת מספיק. נרצה לבצע חלוקה נוספת של הדאטה ל-3 חלקים בפועל, לאמן על חלק האימון, לבחור מודל לפי התוצאות על הולידיישן, ואז להשתמש בטסט בשביל הערכת השגיאה האמיתית של המודל שלנו
6. במטרה להסתכל על הפרויקט כמשימה של דאטה סיינטיסט, אני מעוניין "להלביש" על גבי המודל מתודה שתנגיש את השימוש בפלט שלו. לדוג', הנגשה ויזואלית של דירוגים ממוצעים של המנוע את המשתמשים השונים, ממשק שמאפשר למשתמש להכניס מוצר ולקבל את הדירוג של המנוע עליו / המלצות של המנוע עבור אותו המשתמש, וכדומה. דורש מחקר לגבי ויזואליזציות והנגשות רלוונטיות בעולם של מערכות המלצה.

**הערה לגבי סעיף 3**: יש לבצע מחקר לגבי יכולת השימוש בפיצ'רים על המוצרים / המשתמשים עצמם, שכן האלגוריתמים הקיימים לא יודעים להשתמש בצורה אקטיבית בנתוני "מסגרת" נוספים על המשתמשים ועל המוצרים.

ביציאה לפרויקט אני מניח שאוכל למצוא דרכים להשתמש בנתונים אלו בתהליך האימון, אך ייתכן שבתהליך המחקר שאעשה על המימושים הקיימים כיום, אגיע למסקנות אחרות לגבי יכולת השימוש בהם, ואעבור להשתמש בעיקר / רק בנתוני הדירוג עצמם 